lowranceusr.cc

```diff
@@ -423,28 +975,399 @@ lowranceusr_parse_waypt(Waypoint* wpt_tmp)
+  }
+
+  if (reading_version > 4) {
+    /* USR 5 and 6 have a second Unit Number captured in Waypoints */
+    fsdata->uid_unit2 = gbfgetint32(file_in);
+  }
+
+  /* Long/Lat */
+  wpt_tmp->longitude = lon_mm_to_deg(gbfgetint32(file_in));
+  wpt_tmp->latitude = lat_mm_to_deg(gbfgetint32(file_in));
+
+  /* Flags, discard for now */
+  fsdata->flags = gbfgetint32(file_in);
+
+  /* Icon ID */
+  fsdata->icon_num = gbfgetint16(file_in);
+  wpt_tmp->icon_descr = lowranceusr4_find_desc_from_icon_number(fsdata->icon_num);
+  if (wpt_tmp->icon_descr.isNull()) {
+    char nbuf[10];
+    snprintf(nbuf, sizeof(nbuf), "%d", le_read32(name_buff));
+    wpt_tmp->icon_descr = nbuf;
+  }
+
+  /* Color ID */
+  fsdata->color = gbfgetint16(file_in);
+  fsdata->color_desc = lowranceusr4_find_color_from_icon_number_plus_color_index(fsdata->icon_num, fsdata->color);
+
+  /* Waypoint descr; input is 2 bytes per char, we convert to 1 */
+  int desc_len = lowranceusr4_readstr(&desc_buff[0], MAXUSRSTRINGSIZE, file_in, 2);
+  if (desc_len) {
+    desc_buff[desc_len] = '\0';
+    wpt_tmp->description = desc_buff;
+  }
+
+  /* Alarm radius; XXX: I'm not sure what the units are here,
+     assuming meters but may be feet? */
+  WAYPT_SET(wpt_tmp, proximity, gbfgetflt(file_in));
+
+  /* Creation date/time; the date is a Julian day number, and the
+     time is a unix timestamp. */
+  uint create_date = gbfgetint32(file_in);
+  uint create_time = gbfgetint32(file_in);
+
+  // Julian date 2440487 is 1/1/1970.  If that's the date we're working
+  // with, as a practical matter, we have no date, so don't even compute
+  // or set it.
+  if (create_date > 2440587) {
+    wpt_tmp->SetCreationTime(lowranceusr4_get_timestamp(create_date,
+                             create_time));
+  }
+
+  /* Unused byte */
+  char unused_byte = gbfgetc(file_in);
+
+  /* Altitude/Depth in feet */
+  wpt_tmp->altitude = FEET_TO_METERS(gbfgetflt(file_in));
+
+  /* Loran data, discard for now */
+  int loran_GRI = gbfgetint32(file_in);
+  int loran_Tda = gbfgetint32(file_in);
+  int loran_Tdb = gbfgetint32(file_in);
+
+  if (global_opts.debug_level > 1) {
+    if (global_opts.debug_level == 99) {
+      printf(MYNAME " parse_waypoints: ");
+      if (reading_version > 4) {
+        printf("%08x %08x %08x %08x ",
+               fsdata->UUID1, fsdata->UUID2, fsdata->UUID3, fsdata->UUID4);
+      }
+      printf(" %10u %8d %8d %8d %6d %16s",
+             fsdata->uid_unit, fsdata->uid_seq_low, fsdata->uid_seq_high,
+             waypoint_version, name_len, name_buff);
+      if (reading_version > 4) {
+        printf("  %10u ", fsdata->uid_unit2);
+      }
+      printf(" %+15.10f %+15.10f", wpt_tmp->longitude, wpt_tmp->latitude);
+      printf(" %08x %4d %4d %7s", fsdata->flags, fsdata->icon_num, fsdata->color, qPrintable(fsdata->color_desc));
+      printf(" %6d %16s", desc_len, desc_buff);
+      QDateTime dt;
+      QTime t;
+      dt = wpt_tmp->GetCreationTime();
```

Comment:
Changing all the printfs to qDebugs gets the variable creation and manipulation out of this nested if debug/if more debuglevel.